File: DeenTime/backend/DeenTime.Api/Controllers/IslamicContentController.cs
using DeenTime.Api.Services.IslamicContent;
using DeenTime.Infrastructure;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace DeenTime.Api.Controllers;

[ApiController]
[Authorize("SuperUser")]
[Route("api/v1/islamic-content")]
public sealed class IslamicContentController(
    IDbContextFactory<AppDbContext> dbFactory,
    IIslamicContentSyncQueue syncQueue,
    HadithProviderClient hadithClient) : ControllerBase
{
    [HttpGet("summary")]
    public async Task<IActionResult> Summary(CancellationToken cancellationToken)
    {
        await using var db = await dbFactory.CreateDbContextAsync(cancellationToken);

        var quranEditionCount = await db.QuranEditions.CountAsync(cancellationToken);
        var quranTextCount = await db.QuranEditions.CountAsync(
            edition => edition.Format == "text", cancellationToken);
        var quranAudioCount = await db.QuranEditions.CountAsync(
            edition => edition.Format == "audio", cancellationToken);
        var quranLanguageCount = await db.QuranEditions.Select(edition => edition.Language)
            .Distinct().CountAsync(cancellationToken);
        var quranCachedPayloads = await db.IslamicContentCacheEntries.CountAsync(
            entry => entry.Provider == QuranProviderClient.ProviderName, cancellationToken);
        var quranCachedBytes = await db.IslamicContentCacheEntries
            .Where(entry => entry.Provider == QuranProviderClient.ProviderName)
            .Select(entry => (long?)entry.PayloadBytes)
            .SumAsync(cancellationToken) ?? 0;

        var hadithBookCount = await db.HadithBooks.CountAsync(cancellationToken);
        var hadithChapterCount = await db.HadithChapters.CountAsync(cancellationToken);
        var hadithRecordCount = await db.HadithRecords.CountAsync(cancellationToken);
        var hadithArabicCount = await db.HadithRecords.CountAsync(
            record => !string.IsNullOrEmpty(record.HadithArabic), cancellationToken);
        var hadithEnglishCount = await db.HadithRecords.CountAsync(
            record => !string.IsNullOrEmpty(record.HadithEnglish), cancellationToken);
        var hadithUrduCount = await db.HadithRecords.CountAsync(
            record => !string.IsNullOrEmpty(record.HadithUrdu), cancellationToken);
        var states = await db.IslamicContentSyncStates.AsNoTracking()
            .OrderBy(state => state.Provider)
            .ToArrayAsync(cancellationToken);

        return Ok(new
        {
            quran = new
            {
                editionCount = quranEditionCount,
                textEditionCount = quranTextCount,
                audioEditionCount = quranAudioCount,
                languageCount = quranLanguageCount,
                cachedPayloads = quranCachedPayloads,
                cachedBytes = quranCachedBytes,
                upstreamServer = IslamicContentOptions.RequiredQuranBaseUrl.TrimEnd('/'),
                endpointCount = QuranProviderClient.EndpointTemplates.Length,
                endpointTemplates = QuranProviderClient.EndpointTemplates
            },
            hadith = new
            {
                configured = hadithClient.IsConfigured,
                bookCount = hadithBookCount,
                chapterCount = hadithChapterCount,
                recordCount = hadithRecordCount,
                languages = HadithProviderClient.Languages,
                languageCoverage = new
                {
                    ar = hadithArabicCount,
                    en = hadithEnglishCount,
                    ur = hadithUrduCount
                }
            },
            qibla = new
            {
                provider = QiblaProviderClient.ProviderName,
                providerOrganization = QiblaProviderClient.ProviderOrganization,
                upstreamServer = IslamicContentOptions.RequiredAlAdhanBaseUrl.TrimEnd('/'),
                endpointCount = 2,
                endpointTemplates = new[]
                {
                    QiblaProviderClient.DirectionEndpointTemplate,
                    QiblaProviderClient.CompassEndpointTemplate
                },
                responseFormats = new[] { "application/json", "image/png" },
                metadata = "/public/content/qibla/metadata"
            },
            publicApi = new
            {
                capabilities = "/public/content/capabilities",
                quranBase = "/public/content/quran",
                hadithBase = "/public/content/hadith",
                qiblaBase = "/public/content/qibla"
            },
            syncStates = states
        });
    }

    [HttpGet("quran/editions")]
    public async Task<IActionResult> QuranEditions(
        [FromQuery] string? language,
        [FromQuery] string? format,
        [FromQuery] string? type,
        CancellationToken cancellationToken)
    {
        await using var db = await dbFactory.CreateDbContextAsync(cancellationToken);
        var query = db.QuranEditions.AsNoTracking();
        if (!string.IsNullOrWhiteSpace(language)) query = query.Where(edition => edition.Language == language);
        if (!string.IsNullOrWhiteSpace(format)) query = query.Where(edition => edition.Format == format);
        if (!string.IsNullOrWhiteSpace(type)) query = query.Where(edition => edition.Type == type);

        var editions = await query.OrderBy(edition => edition.Language)
            .ThenBy(edition => edition.EnglishName)
            .ToArrayAsync(cancellationToken);
        return Ok(new { data = editions, total = editions.Length });
    }

    [HttpGet("status")]
    public async Task<IActionResult> Status(CancellationToken cancellationToken)
    {
        await using var db = await dbFactory.CreateDbContextAsync(cancellationToken);
        return Ok(await db.IslamicContentSyncStates.AsNoTracking()
            .OrderBy(state => state.Provider)
            .ToArrayAsync(cancellationToken));
    }

    [HttpPost("sync/quran")]
    public IActionResult SyncQuran([FromBody] QuranSyncRequest request)
    {
        var scope = request.Scope.Trim().ToLowerInvariant();
        if (scope is not ("catalog" or "text" or "all"))
            return BadRequest(new { error = "Scope must be catalog, text, or all." });
        if (!syncQueue.TryQueue("quran", scope))
            return Conflict(new { error = "A Qur'an synchronization is already queued or running." });
        return Accepted(new { provider = "quran", scope, status = "queued" });
    }

    [HttpPost("sync/hadith")]
    public IActionResult SyncHadith()
    {
        if (!hadithClient.IsConfigured)
            return Problem(
                statusCode: StatusCodes.Status503ServiceUnavailable,
                title: "Hadith provider is not configured",
                detail: "Configure the server-side Hadith API key before starting an import.");
        if (!syncQueue.TryQueue("hadith", "all"))
            return Conflict(new { error = "A Hadith synchronization is already queued or running." });
        return Accepted(new { provider = "hadith", scope = "all", status = "queued" });
    }
}

public sealed record QuranSyncRequest(string Scope = "catalog");

File: DeenTime/backend-node/src/modules/islamic-content/islamic-content.controller.ts
import { Body, Controller, Get, HttpCode, Post, Query } from '@nestjs/common';
import { z } from 'zod';
import { Authorize } from '../../common/auth/authorize.decorator.js';
import { badRequest, conflict, problem, validationProblem } from '../../common/errors.js';
import { validated } from '../../common/zod-validation.pipe.js';
import type { Prisma } from '../../generated/prisma/client.js';
import { PrismaService } from '../../prisma/prisma.service.js';
import { HadithProviderClient } from '../../integrations/islamic-content/hadith-provider.client.js';
import { IslamicContentOptions } from '../../integrations/islamic-content/islamic-content-options.js';
import { QiblaProviderClient } from '../../integrations/islamic-content/qibla-provider.client.js';
import { QuranProviderClient } from '../../integrations/islamic-content/quran-provider.client.js';
import { QURAN_SYNC_SCOPES } from './islamic-content-sync.service.js';
import { IslamicContentSyncQueue } from './sync-queue.js';
import { toSyncStateJson } from './sync-state.js';

/** QuranSyncRequest(string Scope = "catalog"). */
const quranSyncSchema = z.object({ scope: z.string().nullable().optional() });

function firstQueryValue(query: Record<string, unknown>, name: string): string | undefined {
  const key = Object.keys(query).find((candidate) => candidate.toLowerCase() === name.toLowerCase());
  const value = key === undefined ? undefined : query[key];
  const single = Array.isArray(value) ? value[0] : value;
  return typeof single === 'string' ? single : undefined;
}

/** Predicate for `!string.IsNullOrEmpty(column)`. */
function nonEmpty(field: 'hadithArabic' | 'hadithEnglish' | 'hadithUrdu'): Prisma.HadithRecordWhereInput {
  return { AND: [{ [field]: { not: null } }, { [field]: { not: '' } }] };
}

@Controller('api/v1/islamic-content')
@Authorize('SuperUser')
export class IslamicContentController {
  constructor(
    private readonly prisma: PrismaService,
    private readonly syncQueue: IslamicContentSyncQueue,
    private readonly hadithClient: HadithProviderClient,
  ) {}

  @Get('summary')
  async summary() {
    const quranProvider = QuranProviderClient.providerName;
    const [
      editionCount,
      textEditionCount,
      audioEditionCount,
      languages,
      cachedPayloads,
      cachedBytes,
      bookCount,
      chapterCount,
      recordCount,
      arabicCount,
      englishCount,
      urduCount,
      states,
    ] = await Promise.all([
      this.prisma.quranEdition.count(),
      this.prisma.quranEdition.count({ where: { format: 'text' } }),
      this.prisma.quranEdition.count({ where: { format: 'audio' } }),
      this.prisma.quranEdition.findMany({ distinct: ['language'], select: { language: true } }),
      this.prisma.islamicContentCacheEntry.count({ where: { provider: quranProvider } }),
      this.prisma.islamicContentCacheEntry.aggregate({ _sum: { payloadBytes: true }, where: { provider: quranProvider } }),
      this.prisma.hadithBook.count(),
      this.prisma.hadithChapter.count(),
      this.prisma.hadithRecord.count(),
      this.prisma.hadithRecord.count({ where: nonEmpty('hadithArabic') }),
      this.prisma.hadithRecord.count({ where: nonEmpty('hadithEnglish') }),
      this.prisma.hadithRecord.count({ where: nonEmpty('hadithUrdu') }),
      this.prisma.islamicContentSyncState.findMany({ orderBy: { provider: 'asc' } }),
    ]);

    return {
      quran: {
        editionCount,
        textEditionCount,
        audioEditionCount,
        languageCount: languages.length,
        cachedPayloads,
        cachedBytes: Number(cachedBytes._sum.payloadBytes ?? 0n),
        upstreamServer: IslamicContentOptions.requiredQuranBaseUrl.replace(/\/+$/, ''),
        endpointCount: QuranProviderClient.endpointTemplates.length,
        endpointTemplates: QuranProviderClient.endpointTemplates,
      },
      hadith: {
        configured: this.hadithClient.isConfigured,
        bookCount,
        chapterCount,
        recordCount,
        languages: HadithProviderClient.languages,
        languageCoverage: { ar: arabicCount, en: englishCount, ur: urduCount },
      },
      qibla: {
        provider: QiblaProviderClient.providerName,
        providerOrganization: QiblaProviderClient.providerOrganization,
        upstreamServer: IslamicContentOptions.requiredAlAdhanBaseUrl.replace(/\/+$/, ''),
        endpointCount: 2,
        endpointTemplates: [QiblaProviderClient.directionEndpointTemplate, QiblaProviderClient.compassEndpointTemplate],
        responseFormats: ['application/json', 'image/png'],
        metadata: '/public/content/qibla/metadata',
      },
      publicApi: {
        capabilities: '/public/content/capabilities',
        quranBase: '/public/content/quran',
        hadithBase: '/public/content/hadith',
        qiblaBase: '/public/content/qibla',
      },
      syncStates: states.map(toSyncStateJson),
    };
  }

  @Get('quran/editions')
  async quranEditions(@Query() query: Record<string, unknown>) {
    const language = firstQueryValue(query, 'language');
    const format = firstQueryValue(query, 'format');
    const type = firstQueryValue(query, 'type');
    const where: Prisma.QuranEditionWhereInput = {};
    if (language?.trim()) where.language = language;
    if (format?.trim()) where.format = format;
    if (type?.trim()) where.type = type;

    const editions = await this.prisma.quranEdition.findMany({ where, orderBy: [{ language: 'asc' }, { englishName: 'asc' }] });
    const data = editions.map((edition) => ({
      identifier: edition.identifier,
      language: edition.language,
      name: edition.name,
      englishName: edition.englishName,
      format: edition.format,
      type: edition.type,
      direction: edition.direction,
      syncedAtUtc: edition.syncedAtUtc,
    }));
    return { data, total: data.length };
  }

  @Get('status')
  async status() {
    const states = await this.prisma.islamicContentSyncState.findMany({ orderBy: { provider: 'asc' } });
    return states.map(toSyncStateJson);
  }

  @Post('sync/quran')
  @HttpCode(202)
  syncQuran(@Body(validated(quranSyncSchema)) body: z.infer<typeof quranSyncSchema>) {
    if (body.scope === null) throw validationProblem({ Scope: ['The Scope field is required.'] });
    const scope = (body.scope ?? 'catalog').trim().toLowerCase();
    if (!QURAN_SYNC_SCOPES.includes(scope)) throw badRequest({ error: 'Scope must be catalog, text, or all.' });
    if (!this.syncQueue.tryQueue('quran', scope)) throw conflict({ error: "A Qur'an synchronization is already queued or running." });
    return { provider: 'quran', scope, status: 'queued' };
  }

  @Post('sync/hadith')
  @HttpCode(202)
  syncHadith() {
    if (!this.hadithClient.isConfigured) {
      throw problem(503, 'Hadith provider is not configured', 'Configure the server-side Hadith API key before starting an import.');
    }
    if (!this.syncQueue.tryQueue('hadith', 'all')) throw conflict({ error: 'A Hadith synchronization is already queued or running.' });
    return { provider: 'hadith', scope: 'all', status: 'queued' };
  }
}
